File: examples/mel_spectrogram.rs
use non_empty_slice::NonEmptyVec;
/// Mel-frequency spectrogram example
///
/// This example demonstrates:
/// - Computing mel-frequency spectrograms
/// - Using logarithmic (dB) amplitude scaling
/// - Accessing mel-frequency axes
use spectrograms::{
    LogParams, MelDbSpectrogram, MelParams, SpectrogramParams, StftParams, WindowType, nzu,
};
use std::f64::consts::PI;

fn main() -> Result<(), Box<dyn std::error::Error>> {
    // Generate a more complex signal: two sine waves
    let sample_rate = 16000.0;
    let duration = 1.0;

    let samples: Vec<f64> = (0..(duration * sample_rate) as usize)
        .map(|i| {
            let t = i as f64 / sample_rate;
            // 440 Hz (A4) + 880 Hz (A5)
            (2.0 * PI * 440.0 * t).sin() + 0.5 * (2.0 * PI * 880.0 * t).sin()
        })
        .collect();
    let samples = NonEmptyVec::new(samples).unwrap();
    println!("Generated {} samples with two frequencies", samples.len());

    // Set up STFT parameters
    let stft = StftParams::new(nzu!(512), nzu!(256), WindowType::Hanning, true)?;
    let params = SpectrogramParams::new(stft, sample_rate)?;

    // Set up mel filterbank parameters
    let mel = MelParams::new(
        nzu!(80), // n_mels: number of mel bands
        0.0,      // f_min: minimum frequency (Hz)
        8000.0,   // f_max: maximum frequency (Hz)
    )?;

    // Set up logarithmic (dB) scaling
    let db = LogParams::new(-80.0)?; // floor at -80 dB

    // Compute mel spectrogram in dB scale
    let spec = MelDbSpectrogram::compute(&samples, &params, &mel, Some(&db))?;

    println!("\nMel Spectrogram properties:");
    println!("  Mel bands: {}", spec.n_bins());
    println!("  Time frames: {}", spec.n_frames());

    // Access mel-frequency axis
    let mel_freqs = spec.axes().frequencies();
    println!("\nMel frequency range:");
    println!("  Lowest mel center: {:.1} Hz", mel_freqs[0]);
    println!(
        "  Highest mel center: {:.1} Hz",
        mel_freqs[mel_freqs.len().get() - 1]
    );

    // Print a few mel band center frequencies
    println!("\nFirst 10 mel band centers (Hz):");
    for (i, &freq) in mel_freqs.iter().take(10).enumerate() {
        println!("  Mel band {}: {:.1} Hz", i, freq);
    }

    // Find average energy per mel band (across all frames)
    println!("\nAverage energy per mel band (dB):");
    for i in (0..spec.n_bins().get()).step_by(spec.n_bins().get() / 5) {
        let row = spec.data().row(i);
        let avg = row.iter().sum::<f64>() / row.len() as f64;
        println!("  Mel band {} ({:.1} Hz): {:.2} dB", i, mel_freqs[i], avg);
    }

    Ok(())
}

File: python/examples/mel_spectrogram.py
#!/usr/bin/env python3
"""
Mel Spectrogram Example

Demonstrates computing mel-scale spectrograms with different amplitude scales.
"""

import numpy as np
import spectrograms as sg


def generate_chirp(sample_rate, duration, f_start, f_end):
    """Generate a linear frequency chirp from f_start to f_end."""
    t = np.linspace(0, duration, int(sample_rate * duration), dtype=np.float64)
    # Instantaneous frequency increases linearly
    k = (f_end - f_start) / duration
    phase = 2 * np.pi * (f_start * t + 0.5 * k * t**2)
    return np.sin(phase)


def main():
    print("=" * 60)
    print("Mel Spectrogram Example")
    print("=" * 60)

    # Generate a chirp signal (frequency sweep from 200 Hz to 4000 Hz)
    sample_rate = 16000
    duration = 2.0
    f_start = 200.0
    f_end = 4000.0

    samples = generate_chirp(sample_rate, duration, f_start, f_end)

    print("\nGenerated chirp signal:")
    print(f"  Sample rate: {sample_rate} Hz")
    print(f"  Duration: {duration} s")
    print(f"  Frequency sweep: {f_start} Hz → {f_end} Hz")
    print(f"  Samples: {len(samples)}")

    # Configure parameters
    stft = sg.StftParams(
        n_fft=512, hop_size=256, window=sg.WindowType.hanning, centre=True
    )
    params = sg.SpectrogramParams(stft, sample_rate=sample_rate)

    # Mel filterbank parameters
    mel_params = sg.MelParams(
        n_mels=80,  # Number of mel bands
        f_min=0.0,  # Minimum frequency
        f_max=8000.0,  # Maximum frequency (Nyquist)
    )

    print("\nMel filterbank parameters:")
    print(f"  Number of mel bands: {mel_params.n_mels}")
    print(f"  Frequency range: {mel_params.f_min} - {mel_params.f_max} Hz")

    # 1. Compute mel power spectrogram
    print("\n" + "-" * 60)
    print("1. Mel Power Spectrogram")
    print("-" * 60)

    mel_power = sg.compute_mel_power_spectrogram(samples, params, mel_params)

    print(f"Shape: {mel_power.shape}")
    print(f"Data range: {np.min(mel_power.data):.2e} to {np.max(mel_power.data):.2e}")
    print(f"Mel frequency range: {mel_power.frequency_range()}")

    # 2. Compute mel magnitude spectrogram
    print("\n" + "-" * 60)
    print("2. Mel Magnitude Spectrogram")
    print("-" * 60)

    mel_mag = sg.compute_mel_magnitude_spectrogram(samples, params, mel_params)

    print(f"Shape: {mel_mag.shape}")
    print(f"Data range: {np.min(mel_mag.data):.2e} to {np.max(mel_mag.data):.2e}")

    # 3. Compute mel dB spectrogram (most common for visualization)
    print("\n" + "-" * 60)
    print("3. Mel dB Spectrogram")
    print("-" * 60)

    db_params = sg.LogParams(floor_db=-80.0)  # Clip values below -80 dB
    print(f"dB floor: {db_params.floor_db} dB")

    mel_db = sg.compute_mel_db_spectrogram(samples, params, mel_params, db_params)

    print(f"Shape: {mel_db.shape}")
    print(f"Data range: {np.min(mel_db.data):.2f} to {np.max(mel_db.data):.2f} dB")
    print(f"Duration: {mel_db.duration():.3f} s")

    # Show mel frequency scale (first 10 bands)
    print("\nFirst 10 mel frequency bands:")
    for i in range(min(10, len(mel_db.frequencies))):
        print(f"  Band {i:2d}: {mel_db.frequencies[i]:7.2f} Hz")

    # Compare different scales
    print("\n" + "=" * 60)
    print("Comparison of Amplitude Scales")
    print("=" * 60)

    # Get values at a specific time/frequency point (middle of spectrogram)
    mid_time = mel_power.n_frames // 2
    mid_freq = mel_power.n_bins // 2

    power_val = mel_power.data[mid_freq, mid_time]
    mag_val = mel_mag.data[mid_freq, mid_time]
    db_val = mel_db.data[mid_freq, mid_time]

    print(f"\nValues at bin {mid_freq}, frame {mid_time}:")
    print(f"  Power:     {power_val:.6e}")
    print(f"  Magnitude: {mag_val:.6e}")
    print(f"  Decibels:  {db_val:.2f} dB")

    # Verify relationship: magnitude = sqrt(power)
    print("\nVerifying relationships:")
    print(f"  sqrt(power) = {np.sqrt(power_val):.6e}")
    print(f"  magnitude   = {mag_val:.6e}")
    print(f"  Match: {np.allclose(np.sqrt(power_val), mag_val)}")

    print("\nMel spectrogram example completed!")


if __name__ == "__main__":
    main()
